S/SwiftShader/build_tarballs.jl

```diff
@@ -2,37 +2,85 @@ using BinaryBuilder
-    ArchiveSource("https://github.com/google/gfbuild-swiftshader/releases/download/github%2Fgoogle%2Fgfbuild-swiftshader%2F$(commit)/gfbuild-swiftshader-$(commit)-Windows_x64_Release.zip",
-                  "22354a01605adb3310736fc85cd1e20039bd4532f428226f8b3ea60d0f2bf853"; unpack_target="x86_64-w64-mingw32"),
+    GitSource(source, commit),
+    DirectorySource("./patches"),
 ]
 
 script = raw"""
-cd $WORKSPACE/srcdir
-mkdir -p "${libdir}"
-mv ${target}/lib/* ${libdir}/.
-rm -rf ${libdir}/*version
-install_license ${target}/OPEN_SOURCE_LICENSES.TXT
+cd swiftshader
+
+# Remove architecture-specific flags.
+atomic_patch -p1 ${WORKSPACE}/srcdir/remove_march.patch
+
+CXX_FLAGS=()
+CMAKE_FLAGS=()
+
+# Release build for best performance
+CMAKE_FLAGS+=(-DCMAKE_BUILD_TYPE=Release)
+
+# Install things into $prefix
+CMAKE_FLAGS+=(-DCMAKE_INSTALL_PREFIX=${prefix})
+
+# Explicitly use our cmake toolchain file
+CMAKE_FLAGS+=(-DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN})
+
+# Work around undefined `cinttypes` printing macros.
+CXX_FLAGS+="-D__STDC_FORMAT_MACROS"
+
+# Do not enable the -Werror flag (builds fine without it).
+CMAKE_FLAGS+=(-DSWIFTSHADER_WARNINGS_AS_ERRORS=FALSE)
+
+# Don't build unit tests (they currently error the build because of yet unindentified libc++ issues).
+CMAKE_FLAGS+=(-DSWIFTSHADER_BUILD_TESTS=FALSE)
+
+# Handle Mac SDK <10.12 errors.
+if [[ "${target}" == *-apple-* ]]; then
+    export MACOSX_DEPLOYMENT_TARGET=10.12
+    CMAKE_FLAGS+=(-DLLVM_ENABLE_LIBCXX=ON)
+fi
+
+if [[ "${target}" == *-mingw* ]]; then
+    find . -type f -exec sed -i 's/include <Windows\.h>/include <windows\.h>/g' {} +
+    CXX_FLAGS+=" -DHAVE_UNISTD"
+fi
+
+cmake -B build -S . -GNinja ${CMAKE_FLAGS[@]} -DCMAKE_CXX_FLAGS="${CXX_FLAGS}"
+ninja -C build -j ${nproc} install
+
+folderName=""
+if [[ "${target}" == *-apple-* ]]; then
+    echo "hello"
+    folderName=Darwin
+elif [[ "${target}" == *-linux-* ]]; then
+    folderName=Linux
+elif [[ "${target}" == *-mingw* ]]; then
+    folderName=Windows
+fi
+
+mv build/$folderName/* ${libdir}
+
+install_license LICENSE.txt
 """
 
 platforms = [
     Platform("x86_64", "linux"; libc="glibc"),
-    Platform("x86_64", "macos"),
-    Platform("x86_64", "windows"),
+    Platform("x86_64", "macos"; os_version = v"10.15"),
+    # TODO: Build on Windows.
+    # Might want to take some inspiration from https://github.com/google/gfbuild-swiftshader/blob/master/build.sh
+    # Platform("x86_64", "windows"),
 ]
 
 products = [
-    LibraryProduct(["libvulkan", "vk_swiftshader"], :libvulkan),
+    LibraryProduct("libvulkan", :libvulkan),
+    LibraryProduct("libvk_swiftshader", :libvk_swiftshader),
 ]
 
 # Dependencies that must be installed before this package can be built
 dependencies = Dependency[]
 
-build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies)
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies, preferred_gcc_version=v"11")
```

Comment:
Please don't use such a late version of GCC.  Use the ***oldest*** version that can actually compile the source code, and in general avoid going past 9 or 10, especially for C++ code, like this one.

Reply:
I started with GCC 4, then 5, then 10 and it did not work; GCC 11 seemed to be required. Fortunately, this library should be rarely added as a dependency and is more of a "final product".
I am not 100% certain that the errors encountered while experimenting with older GCC versions could not be worked around in some way (e.g. patching more things). I can try again if it really is a major pain to include such a late version of GCC.

Reply:
Well, if you really need GCC 11 then be aware that this will likely not work at all with Julia v1.6.

Side note, even if not strictly required, it'd be good to specify the julia compat, like
```suggestion
build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; julia_compat="1.6", preferred_gcc_version=v"11")
```
or whatever

Reply:
> Well, if you really need GCC 11 then be aware that this will likely not work at all with Julia v1.6.

I will go with GCC 11 for now, if it becomes a pain it can always be attempted to require a lower version as an improvement.

Given the suggestion, did you mean Julia < 1.6 or should the compat be e.g. >= 1.7?